File: SkylinkSample/TEMVideoView.m
//
//  TEMVideoView.m
//  SampleApp
//
//  Created by macbookpro on 07/05/2015.
//  Copyright (c) 2015 Temasys Communications. All rights reserved.
//

#import <AVFoundation/AVFoundation.h>

#import "TEMVideoView.h"

@interface TEMVideoView () {
    CGSize renderSize;
}

@property(nonatomic, weak) UIView* renderView;
@property(nonatomic, weak) UIButton* glassButton;

@end

@implementation TEMVideoView

static void init(TEMVideoView *videoView, UIView *renderView) {
    renderView.frame = videoView.bounds;
    renderView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    renderView.translatesAutoresizingMaskIntoConstraints = YES;
    [videoView addSubview:renderView];
    videoView.renderView = renderView;
    
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    button.frame = CGRectMake(0, 0, videoView.bounds.size.width, videoView.bounds.size.height);
    button.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    button.translatesAutoresizingMaskIntoConstraints = YES;
    [button addTarget:videoView action:@selector(videoViewIsTapped:) forControlEvents:UIControlEventTouchUpInside];
    [videoView addSubview:button];
    videoView.glassButton = button;
}

- (id)initWithFrame:(CGRect)frame videoView:(UIView*)renderView {
    self = [super initWithFrame:frame];
    if (self) {
        init(self, renderView);
        renderSize = CGSizeZero;
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    if (_renderView && (renderSize.height > 0 && renderSize.width > 0)) {
        CGSize defaultAspectRatio = CGSizeMake(4, 3);
        CGSize aspectRatio = CGSizeEqualToSize(renderSize, CGSizeZero) ? defaultAspectRatio : renderSize;
        CGRect videoFrame = AVMakeRectWithAspectRatioInsideRect(aspectRatio, self.bounds);
        self.renderView.frame = self.glassButton.frame = videoFrame;
    }
}

#pragma mark - public methods

- (void)layoutSubviews:(CGSize)size {
    renderSize = CGSizeEqualToSize(size, CGSizeZero) ? renderSize : size;
    [self setNeedsLayout];
}

- (UIView*)getRenderSurface {
    return self.renderView;
}

- (UIView*)getTouchSurface {
    return self.glassButton;
}

#pragma mark - IBAction

- (IBAction)videoViewIsTapped:(id)sender {
    if ([self.delegate respondsToSelector:@selector(videoViewIsTapped:)])
        [self.delegate videoViewIsTapped:self];
}

@end

File: SampleAppSwift/SampleAppSwift/TEMVideoView.swift
//
//  TEMVideoView.swift
//  SampleAppSwift
//
//  Created by macbookpro on 08/05/2015.
//  Copyright (c) 2015 Temasys Communications. All rights reserved.
//

import UIKit

protocol TEMVideoViewDelegate {
    
    func videoViewIsTapped(videoView : TEMVideoView)
    
}

class TEMVideoView: UIView {

    var delegate: TEMVideoViewDelegate?
    
    private var renderSize: CGSize?
    
    private weak var renderView: UIView?
    private weak var glassButton: UIButton?
    
    static func initialize(videoView: TEMVideoView, renderView: UIView) {
        renderView.frame = videoView.bounds
        renderView.autoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight
        renderView.setTranslatesAutoresizingMaskIntoConstraints(true)
        videoView.addSubview(renderView)
        videoView.renderView = renderView
        
        var button: UIButton = UIButton.buttonWithType(UIButtonType.Custom) as! UIButton
        button.frame = CGRectMake(0, 0, videoView.bounds.size.width, videoView.bounds.size.height);
        button.autoresizingMask = UIViewAutoresizing.FlexibleWidth | UIViewAutoresizing.FlexibleHeight
        button.setTranslatesAutoresizingMaskIntoConstraints(true)
        button.addTarget(videoView, action: "videoViewIsTapped:", forControlEvents: UIControlEvents.TouchUpInside)
        videoView.addSubview(button)
        videoView.glassButton = button
    }
    
    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    init(frame: CGRect, videoView: UIView) {
        super.init(frame: frame)
        TEMVideoView.initialize(self, renderView: videoView)
        renderSize = CGSizeZero;
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        if (renderView != nil && (renderSize?.height > 0 && renderSize?.width > 0)) {
            let defaultAspectRatio: CGSize = CGSizeMake(4, 3)
            let aspectRatio: CGSize! = CGSizeEqualToSize(renderSize!, CGSizeZero) ? defaultAspectRatio : renderSize
            let videoFrame: CGRect = AVMakeRectWithAspectRatioInsideRect(aspectRatio, self.bounds)
            self.renderView?.frame = videoFrame
            self.glassButton?.frame = videoFrame
        }
    }

    func layoutSubviews(size: CGSize) {
        renderSize = CGSizeEqualToSize(size, CGSizeZero) ? renderSize : size
        self.setNeedsLayout()
    }

    func getRenderSurface() -> UIView {
        return renderView!
    }

    func getTouchSurface() -> UIView {
        return glassButton!
    }
    
    @IBAction func videoViewIsTapped(sender: UIButton) {
        if (self.delegate != nil) {
            self.delegate?.videoViewIsTapped(self)
        }
    }

}
